Remove Enzyme from Arrow Vega Lite Charts

kmcgrady opened the pull request:
## Describe your changes
Remove Enzyme from testing Arrow Vega Lite Charts. This is relatively easy to solve by checking on the canvas drawings match as expected from snapshots (it's a lot). I think it's good for now, but in the future, we can optimize to check for specific calls done correctly.

## Testing Plan

- Unit tests are rewritten. I did test that changes to themes/colors actually break the test. Code Reviewer can just determine if this is a good enough change.

---

**Contribution License Agreement**

By submitting this pull request you agree that all contributions to this project are made under the Apache 2.0 license.

kmcgrady committed:
Remove Enzyme from Arrow Vega Lite Charts

willhuang1997 commented on frontend/lib/src/components/elements/ArrowVegaLiteChart/ArrowVegaLiteChart.test.tsx:
```diff
@@ -60,30 +63,36 @@ const getProps = (props: Partial<PropsWithHeight> = {}): PropsWithHeight => ({
+
 describe("VegaLiteChart Element", () => {
-  it("renders without crashing", () => {
+  it("renders without crashing", async () => {
```
nit: Do we need async here?

kmcgrady replied:
We do. ArrowVegaLite has async activity. See https://github.com/streamlit/streamlit/blob/develop/frontend/lib/src/components/elements/ArrowVegaLiteChart/ArrowVegaLiteChart.tsx#L138

willhuang1997 commented on frontend/lib/src/components/elements/ArrowVegaLiteChart/ArrowVegaLiteChart.test.tsx:
```diff
@@ -60,30 +63,36 @@ const getProps = (props: Partial<PropsWithHeight> = {}): PropsWithHeight => ({
-    const wrapper = mount(<ArrowVegaLiteChart {...props} />)
 
-    expect(wrapper.find("StyledVegaLiteChartContainer").length).toBe(1)
+    await render(<ArrowVegaLiteChart {...props} />)
```
nit: If async here not needed, we can remove this await i think

kmcgrady replied:
Ah I should probably just do the `await new Promise(process.nextTick), which allows the async operation to complete. Will update.

willhuang1997 commented on frontend/lib/src/components/elements/ArrowVegaLiteChart/ArrowVegaLiteChart.test.tsx:
```diff
@@ -60,30 +63,36 @@ const getProps = (props: Partial<PropsWithHeight> = {}): PropsWithHeight => ({
-    expect(generatedSpec.config.axis.labelColor).toBe(
-      mockTheme.emotion.colors.bodyText
-    )
+    await new Promise(process.nextTick)
```
Is there a reason why we need to await for the next tick? If we don't, maybe need a comment why this is needed?

kmcgrady replied:
I can add a comment. Essentially this waits for any async code to resolve (aka the await statements that create the view).

kmcgrady committed:
Favor snapshot testing over canvas checks

kmcgrady committed:
Remove snapshots

kmcgrady committed:
Remove unnecessary imports

kmcgrady committed:
Add snapshots

kmcgrady commented:
Per suggestion from @LukasMasuch I removed the canvas checks in favor of screenshots. We currently had a streamlit and default theme test, so I added the purple background. We decided this because the canvas checks are cumbersome to understand the differences whereas a screenshot check can show a problem more easily. I hope this doesn't create flakiness, but I think we've seen this work well in the past for vega.

Those tests are in cypress and so I did the migration to playwright. Would love guidance on whether I did the rewrite effectively.

Let me know what you think @willhuang1997

kmcgrady committed:
Update test to remove iterable

lukasmasuch commented on e2e_playwright/st_arrow_vega_lite_chart_test.py:
```diff
@@ -0,0 +1,111 @@
+
+def test_arrow_vega_lite_chart(themed_app: Page):
+    """Tests that it displays charts on the DOM"""
+    arrow_vega_lite_charts = themed_app.locator("[data-testid=stArrowVegaLiteChart]")
```
nit: there is also a direct method to get by test ID:

```suggestion
    arrow_vega_lite_charts = themed_app.get_by_test_id("stArrowVegaLiteChart")
```

lukasmasuch commented on e2e_playwright/st_arrow_vega_lite_chart_test.py:
```diff
@@ -0,0 +1,111 @@
+from e2e_playwright.conftest import ImageCompareFunction
+
+
+def test_arrow_vega_lite_chart(themed_app: Page):
```
nit: since this test isn't really testing any theming-related aspects, it could also just use `app` to only run the test once:

```suggestion
def test_arrow_vega_lite_chart(app: Page):
```

lukasmasuch commented:
I did a quick glance at the playwright test file, and it looks good 👍

willhuang1997 commented on e2e_playwright/st_arrow_vega_lite_chart_test.py:
```diff
@@ -0,0 +1,111 @@
+
+def test_arrow_vega_lite_chart_sets_chart_width(themed_app: Page):
+    """Tests that it sets the correct chart width"""
+    arrow_vega_lite_charts = themed_app.locator("[data-testid=stArrowVegaLiteChart]")
```
nit: `arrow_vega_lite_charts = themed_app.locator("[data-testid=stArrowVegaLiteChart]")` could be
```suggestion
    arrow_vega_lite_charts = themed_app.get_by_test_id("stArrowVegaLiteChart")")
```

willhuang1997 commented on e2e_playwright/st_arrow_vega_lite_chart_test.py:
```diff
@@ -0,0 +1,111 @@
+    themed_app: Page, assert_snapshot: ImageCompareFunction
+):
+    """Tests that it displays interactive charts on the DOM"""
+    arrow_vega_lite_charts = themed_app.locator("[data-testid=stArrowVegaLiteChart]")
```
nit: `arrow_vega_lite_charts = themed_app.locator("[data-testid=stArrowVegaLiteChart]")` could be
```suggestion
    arrow_vega_lite_charts = themed_app.get_by_test_id("stArrowVegaLiteChart")")
```

willhuang1997 commented on e2e_playwright/st_arrow_vega_lite_chart_test.py:
```diff
@@ -0,0 +1,111 @@
+    themed_app: Page, assert_snapshot: ImageCompareFunction
+):
+    """Tests that it displays the same plot in different ways"""
+    arrow_vega_lite_charts = themed_app.locator("[data-testid=stArrowVegaLiteChart]")
```
ditto

willhuang1997 replied:
Just same with all of the other ones in the rest of file.

kmcgrady committed:
Addressed PR comments

kmcgrady committed:
Try change in preloading fonts

kmcgrady committed:
Merge branch 'develop' into refactor/rtl-arrow-vega-lite
